osu.Game/Beatmaps/BeatmapInfoExtensions.cs

```diff
@@ -29,10 +29,29 @@ public static RomanisableString GetDisplayTitleRomanisable(this IBeatmapInfo bea
             return new RomanisableString($"{metadata.GetPreferred(true)}".Trim(), $"{metadata.GetPreferred(false)}".Trim());
         }
 
-        public static string[] GetSearchableTerms(this IBeatmapInfo beatmapInfo) => new[]
+        public static ReadOnlySpan<string> GetSearchableTerms(this IBeatmapInfo beatmapInfo)
         {
-            beatmapInfo.DifficultyName
-        }.Concat(beatmapInfo.Metadata.GetSearchableTerms()).Where(s => !string.IsNullOrEmpty(s)).ToArray();
+            Span<string> terms = new string[8];
+            int i = 0;
+            if (!string.IsNullOrEmpty(beatmapInfo.DifficultyName))
+                terms[i++] = beatmapInfo.DifficultyName;
+            var metadata = beatmapInfo.Metadata;
+            if (!string.IsNullOrEmpty(metadata.Author.Username))
+                terms[i++] = metadata.Author.Username;
+            if (!string.IsNullOrEmpty(metadata.Artist))
+                terms[i++] = metadata.Artist;
+            if (!string.IsNullOrEmpty(metadata.ArtistUnicode))
+                terms[i++] = metadata.ArtistUnicode;
+            if (!string.IsNullOrEmpty(metadata.Title))
+                terms[i++] = metadata.Title;
+            if (!string.IsNullOrEmpty(metadata.TitleUnicode))
+                terms[i++] = metadata.TitleUnicode;
+            if (!string.IsNullOrEmpty(metadata.Source))
+                terms[i++] = metadata.Source;
+            if (!string.IsNullOrEmpty(metadata.Tags))
+                terms[i++] = metadata.Tags;
```

Comment:
This is essentially inlining `BeatmapMetadataInfoExtensions.GetSearchableTerms()`? Why is that method (which is used elsewhere) allowed to be slow?

I also generally disagree with restating the method locally here. There is surely a way to do this better to eliminate duplication without sacrificing performance. Even if via internal methods or something.

Reply:
I can't think of a way to return multiple strings without a bunch of overhead and memory allocations.

Reply:
What I would propose:

```diff
diff --git a/osu.Game/Beatmaps/BeatmapInfoExtensions.cs b/osu.Game/Beatmaps/BeatmapInfoExtensions.cs
index 460944d2fa..7ecb2708d6 100644
--- a/osu.Game/Beatmaps/BeatmapInfoExtensions.cs
+++ b/osu.Game/Beatmaps/BeatmapInfoExtensions.cs
@@ -2,6 +2,8 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System;
+using System.Collections.Generic;
+using Microsoft.Toolkit.HighPerformance;
 using osu.Framework.Localisation;
 
 namespace osu.Game.Beatmaps
@@ -31,24 +33,14 @@ public static RomanisableString GetDisplayTitleRomanisable(this IBeatmapInfo bea
 
         public static ReadOnlySpan<string> GetSearchableTerms(this IBeatmapInfo beatmapInfo)
         {
-            string[] terms = new string[8];
-            int i = 0;
-            var metadata = beatmapInfo.Metadata;
-            addIfNotNull(beatmapInfo.DifficultyName);
-            addIfNotNull(metadata.Author.Username);
-            addIfNotNull(metadata.Artist);
-            addIfNotNull(metadata.ArtistUnicode);
-            addIfNotNull(metadata.Title);
-            addIfNotNull(metadata.TitleUnicode);
-            addIfNotNull(metadata.Source);
-            addIfNotNull(metadata.Tags);
-            return terms.AsSpan(0, i);
-
-            void addIfNotNull(string? s)
-            {
-                if (!string.IsNullOrEmpty(s))
-                    terms[i++] = s;
-            }
+            var termsList = new List<string>(BeatmapMetadataInfoExtensions.MAX_SEARCHABLE_TERM_COUNT + 1);
+
+            if (!string.IsNullOrEmpty(beatmapInfo.DifficultyName))
+                termsList.Add(beatmapInfo.DifficultyName);
+
+            BeatmapMetadataInfoExtensions.CollectSearchableTerms(beatmapInfo.Metadata, termsList);
+
+            return termsList.AsSpan();
         }
 
         private static string getVersionString(IBeatmapInfo beatmapInfo) => string.IsNullOrEmpty(beatmapInfo.DifficultyName) ? string.Empty : $"[{beatmapInfo.DifficultyName}]";
diff --git a/osu.Game/Beatmaps/BeatmapMetadataInfoExtensions.cs b/osu.Game/Beatmaps/BeatmapMetadataInfoExtensions.cs
index 738bdb0839..867495bae2 100644
--- a/osu.Game/Beatmaps/BeatmapMetadataInfoExtensions.cs
+++ b/osu.Game/Beatmaps/BeatmapMetadataInfoExtensions.cs
@@ -3,7 +3,7 @@
 
 #nullable disable
 
-using System.Linq;
+using System.Collections.Generic;
 using osu.Framework.Localisation;
 
 namespace osu.Game.Beatmaps
@@ -13,16 +13,32 @@ public static class BeatmapMetadataInfoExtensions
         /// <summary>
         /// An array of all searchable terms provided in contained metadata.
         /// </summary>
-        public static string[] GetSearchableTerms(this IBeatmapMetadataInfo metadataInfo) => new[]
+        public static string[] GetSearchableTerms(this IBeatmapMetadataInfo metadataInfo)
         {
-            metadataInfo.Author.Username,
-            metadataInfo.Artist,
-            metadataInfo.ArtistUnicode,
-            metadataInfo.Title,
-            metadataInfo.TitleUnicode,
-            metadataInfo.Source,
-            metadataInfo.Tags
-        }.Where(s => !string.IsNullOrEmpty(s)).ToArray();
+            var termsList = new List<string>(MAX_SEARCHABLE_TERM_COUNT);
+            CollectSearchableTerms(metadataInfo, termsList);
+            return termsList.ToArray();
+        }
+
+        internal const int MAX_SEARCHABLE_TERM_COUNT = 7;
+
+        internal static void CollectSearchableTerms(IBeatmapMetadataInfo metadataInfo, IList<string> termsList)
+        {
+            if (!string.IsNullOrEmpty(metadataInfo.Author.Username))
+                termsList.Add(metadataInfo.Author.Username);
+            if (!string.IsNullOrEmpty(metadataInfo.Artist))
+                termsList.Add(metadataInfo.Artist);
+            if (!string.IsNullOrEmpty(metadataInfo.ArtistUnicode))
+                termsList.Add(metadataInfo.ArtistUnicode);
+            if (!string.IsNullOrEmpty(metadataInfo.Title))
+                termsList.Add(metadataInfo.Title);
+            if (!string.IsNullOrEmpty(metadataInfo.TitleUnicode))
+                termsList.Add(metadataInfo.TitleUnicode);
+            if (!string.IsNullOrEmpty(metadataInfo.Source))
+                termsList.Add(metadataInfo.Source);
+            if (!string.IsNullOrEmpty(metadataInfo.Tags))
+                termsList.Add(metadataInfo.Tags);
+        }
 
         /// <summary>
         /// A user-presentable display title representing this metadata.
```

This PR as-is:

|                          Method |      Mean |    Error |   StdDev |   Gen0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|-------:|----------:|
|           CarouselBeatmapFilter |  21.70 ns | 0.145 ns | 0.129 ns |      - |         - |
| CriteriaMatchingSpecificRuleset |  44.23 ns | 0.109 ns | 0.091 ns |      - |         - |
|        CriteriaMatchingRangeMax |  50.82 ns | 0.429 ns | 0.359 ns |      - |         - |
|           CriteriaMatchingTerms | 371.61 ns | 1.396 ns | 1.306 ns | 0.0067 |      88 B |
|         CriteriaMatchingCreator |  68.06 ns | 0.592 ns | 0.525 ns |      - |         - |
|      CriteriaMatchingBeatmapIDs | 299.71 ns | 1.709 ns | 1.599 ns | 0.0067 |      88 B |

With the above patch:

|                          Method |      Mean |    Error |   StdDev |   Gen0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|-------:|----------:|
|           CarouselBeatmapFilter |  21.73 ns | 0.250 ns | 0.209 ns |      - |         - |
| CriteriaMatchingSpecificRuleset |  43.59 ns | 0.197 ns | 0.184 ns |      - |         - |
|        CriteriaMatchingRangeMax |  51.44 ns | 0.305 ns | 0.271 ns |      - |         - |
|           CriteriaMatchingTerms | 453.05 ns | 8.648 ns | 8.089 ns | 0.0095 |     120 B |
|         CriteriaMatchingCreator |  62.87 ns | 0.269 ns | 0.238 ns |      - |         - |
|      CriteriaMatchingBeatmapIDs | 379.27 ns | 1.827 ns | 1.620 ns | 0.0095 |     120 B |

No significant extra overhead, significant (in my view) maintainability gain.

Reply:
Okay I tried your changes. I figured the performance overhead was pretty big, so I tried adding the `addIfNotNull` local method to it and that actually improved the performance almost to back where it was before your patch. This way we have the best of both worlds.

With your patch:
|                          Method |      Mean |    Error |   StdDev |   Gen0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|-------:|----------:|
|           CarouselBeatmapFilter |  32.31 ns | 0.444 ns | 0.415 ns |      - |         - |
| CriteriaMatchingSpecificRuleset |  58.79 ns | 0.926 ns | 0.821 ns |      - |         - |
|        CriteriaMatchingRangeMax |  70.23 ns | 0.576 ns | 0.539 ns |      - |         - |
|           CriteriaMatchingTerms | 708.49 ns | 3.453 ns | 2.884 ns | 0.0143 |     120 B |
|         CriteriaMatchingCreator |  94.28 ns | 0.714 ns | 0.668 ns |      - |         - |
|      CriteriaMatchingBeatmapIDs | 602.20 ns | 6.165 ns | 5.767 ns | 0.0143 |     120 B |

With your patch and `addIfNotNull`:
|                          Method |      Mean |    Error |   StdDev |   Gen0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|-------:|----------:|
|           CarouselBeatmapFilter |  31.96 ns | 0.480 ns | 0.449 ns |      - |         - |
| CriteriaMatchingSpecificRuleset |  58.89 ns | 0.601 ns | 0.562 ns |      - |         - |
|        CriteriaMatchingRangeMax |  69.59 ns | 0.641 ns | 0.600 ns |      - |         - |
|           CriteriaMatchingTerms | 650.33 ns | 1.974 ns | 1.846 ns | 0.0143 |     120 B |
|         CriteriaMatchingCreator |  95.07 ns | 1.916 ns | 1.968 ns |      - |         - |
|      CriteriaMatchingBeatmapIDs | 517.26 ns | 2.006 ns | 1.876 ns | 0.0143 |     120 B |

Reply:
Really strange to me that the *local method* is saving time here. But hey, who am I to argue against cold hard data...